README
NAME
    Sub::Uplevel - apparently run a function in a higher stack frame

VERSION
    This documentation describes version 0.21_01

SYNOPSIS
      use Sub::Uplevel;

      sub foo {
          print join " - ", caller;
      }

      sub bar {
          uplevel 1, \&foo;
      }

      #line 11
      bar();    # main - foo.plx - 11

DESCRIPTION
    Like Tcl's uplevel() function, but not quite so dangerous. The idea is
    just to fool caller(). All the really naughty bits of Tcl's uplevel()
    are avoided.

    THIS IS NOT THE SORT OF THING YOU WANT TO DO EVERYDAY

    uplevel
          uplevel $num_frames, \&func, @args;

        Makes the given function think it's being executed $num_frames
        higher than the current stack level. So when they use
        caller($frames) it will actually give caller($frames + $num_frames)
        for them.

        "uplevel(1, \&some_func, @_)" is effectively "goto &some_func" but
        you don't immediately exit the current subroutine. So while you
        can't do this:

            sub wrapper {
                print "Before\n";
                goto &some_func;
                print "After\n";
            }

        you can do this:

            sub wrapper {
                print "Before\n";
                my @out = uplevel 1, &some_func;
                print "After\n";
                return @out;
            }

        "uplevel" will issue a warning if $num_frames is more than the
        current call stack depth.

EXAMPLE
    The main reason I wrote this module is so I could write wrappers around
    functions and they wouldn't be aware they've been wrapped.

        use Sub::Uplevel;

        my $original_foo = \&foo;

        *foo = sub {
            my @output = uplevel 1, $original_foo;
            print "foo() returned:  @output";
            return @output;
        };

    If this code frightens you you should not use this module.

BUGS and CAVEATS
    Well, the bad news is uplevel() is about 5 times slower than a normal
    function call. XS implementation anyone? It also slows down every
    invocation of caller(), regardless of whether uplevel() is in effect.

    Sub::Uplevel overrides CORE::GLOBAL::caller temporarily for the scope of
    each uplevel call. It does its best to work with any previously existing
    CORE::GLOBAL::caller (both when Sub::Uplevel is first loaded and within
    each uplevel call) such as from Contextual::Return or Hook::LexWrap.

    However, if you are routinely using multiple modules that override
    CORE::GLOBAL::caller, you are probably asking for trouble.

    You should load Sub::Uplevel as early as possible within your program.
    As with all CORE::GLOBAL overloading, the overload will not affect
    modules that have already been compiled prior to the overload. One
    module that often is unavoidably loaded prior to Sub::Uplevel is
    Exporter. To forceably recompile Exporter (and Exporter::Heavy) after
    loading Sub::Uplevel, use it with the ":aggressive" tag:

        use Sub::Uplevel qw/:aggressive/;

    The private function "Sub::Uplevel::_force_reload()" may be passed a
    list of additional modules to reload if ":aggressive" is not aggressive
    enough. Reloading modules may break things, so only use this as a last
    resort.

    As of version 0.20, Sub::Uplevel requires Perl 5.6 or greater.

HISTORY
    Those who do not learn from HISTORY are doomed to repeat it.

    The lesson here is simple: Don't sit next to a Tcl programmer at the
    dinner table.

THANKS
    Thanks to Brent Welch, Damian Conway and Robin Houston.

AUTHORS
    David A Golden <dagolden@cpan.org> (current maintainer)

    Michael G Schwern <schwern@pobox.com> (original author)

LICENSE
    Original code Copyright (c) 2001 to 2007 by Michael G Schwern.
    Additional code Copyright (c) 2006 to 2008 by David A Golden.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://www.perl.com/perl/misc/Artistic.html

SEE ALSO
    PadWalker (for the similar idea with lexicals), Hook::LexWrap, Tcl's
    uplevel() at http://www.scriptics.com/man/tcl8.4/TclCmd/uplevel.htm



Changes
Changes for Sub::Uplevel

{{$NEXT}}

0.2800    2017-04-01 13:31:22-04:00 America/New_York

    - No changes from 0.2700-TRIAL

0.2700    2017-03-30 10:09:46-04:00 America/New_York (TRIAL RELEASE)

    [Fixed]

    - Tests now work if '.' is not in @INC

0.2600    2016-08-05 10:46:37-04:00 America/New_York

    - No changes from 0.2501-TRIAL

0.2501    2016-07-29 16:18:45-04:00 America/New_York (TRIAL RELEASE)

    [~Internal~]

    - Optimized calls to caller()

0.25      2015-01-26 20:32:52-05:00 America/New_York

    [Fixed]

    - fixed: 00-compile.t failures under Windows (bug #98230).
      The 00-compile.t file has been moved to a release test and is
      no longer shipped.  [Michael Gray]

    [Meta]

    - moved bug tracker to Github

    - updated repo files explaining how to contribute

    - enabled Travis CI

0.24      2012-02-20 22:18:46 EST5EDT

    - no changes from 0.23_03

0.23_03   2012-02-07 17:16:49 EST5EDT

    - fixed: additional test regex fixed

0.23_02   2012-02-07 14:32:53 EST5EDT

    - fixed: Perl 5.15 series Carp changes output format of messages (adds
      a period).  This fixes the test regular expressions. [Andreas Koenig]

0.23_01   2012-02-02 10:41:31 EST5EDT

    - added: the $Sub::Uplevel::CHECK_FRAMES global constant to enable the
      frame depth checking and compile it out if not needed.  [Adam
      Kennedy]

0.22 Tue Nov 17 05:38:42 EST 2009

    - meta: added repository to metadata [Alexandr Ciornii]

0.21_01 Sat Nov 29 15:24:49 EST 2008

    - added: the ":aggressive" tag may be used to force reloading of 
      Exporter, which often binds caller() prior to Sub::Uplevel's global
      override

0.2002 Thu Sep 11 14:33:09 EDT 2008

    - changed: removed Exporter dependency

    - test fix: fixed strange t/07_uplevel_too_high.t fail on Win32
    
0.2001 Tue Sep  9 22:22:40 EDT 2008

    - test fix: changed prior override test to be more robust (SCHWERN)

0.20 Tue Sep  9 19:23:35 EDT 2008

    - changed: bumped perl requirement to 5.006 and stopped using vars
      (fixes a test bug under Test::More > 0.80)

0.19_03 Fri Jul  4 13:31:21 EDT 2008

    - fixed: load Carp only as needed (fixes problem on 5.005)

0.19_02 Thu Feb 21 14:58:46 EST 2008

    - added: uplevel will warn if uplevel request is more than the call
      stack depth

0.1901 Thu Feb 14 14:07:37 EST 2008

    - fixed: hides the "DB" package from indexers (DB used to support
      @DB::args)
    
0.19 Thu Feb 14 11:50:16 EST 2008

    - bug fix: replacement caller() correctly sets @DB::args if calling
      package is DB

    - repackaging: custom Build.PL, moved all author tests to xt/, added
      LICENSE and INSTALL files

0.18 Wed Oct 31 06:56:13 EDT 2007
    - release version of 0.17_01 changes

0.17_01
    - removed 5.006-specific language and tested on prerelease 5.00505

0.16 Mon Jul 30 09:54:41 EDT 2007
    - release version of 0.15_01 changes

0.15_01 Thu Jul  5 22:54:08 EDT 2007
    - Won't override any existing CORE::GLOBAL::caller when loaded 
    - Localize global caller() override to the scope of the uplevel() call 
      so it can play nicer with things like Contextual::Return and
      Hook::LexWrap that also override caller()

0.14 Sun Nov  5 23:38:46 EST 2006
    - fixed t/99_pod_coverage.t bug
    - added examples directory

0.13 Thu Jun 22 19:47:26 EDT 2006
    - fixed bug in Uplevel.t that caused test failure on FreeBSD for 5.8.0
    - fixed bug in Uplevel.t that caused test failure on bleadperl-5.9.4
    - removed ancient Test::More provided in t/lib
    - switched ok(eq_array()) tests to is_deeper() for better diagnostics
    - added pod/pod coverage checks
    - numbered tests

0.12 Fri May 12 18:33:40 EDT 2006
    - official release of the uplevel stack patch (fixes RT#13893)
    - added Build.PL and switched to boilerplate generated Makefile.PL
    - removed SIGNATURE due to recently discovered Module::Signature
      issues with newline handling and sub-key compatible keyservers
    - updated/added various meta files

0.11_01 Fri Apr 21 00:49:51 EDT 2006
    - uplevel now keeps a proper stack of uplevel calls allowing
      nesting of uplevel and non-uplevel calls

0.10 Thu Apr 20 19:15:20 EDT 2006
    - Stopped warnings about "undefined" on Perl 5.8.8
    - DAGOLDEN added as co-maintainer

0.09  Wed Jul  7 14:52:08 EDT 2004
    - Ok, ok.  I'll put a license on this.

0.08  Wed Oct 22 09:02:38 PDT 2003
    - New die_check.t test was written in a non-portable manner.
      [Thanks Martin Thurn and cpantesters]

0.07  Tue Mar 18 03:03:22 GMT 2003
    - Fixed a test bug due to 5.6.0's differing Carp::croak stack output

0.06  Thu Sep 20 08:50:30 EDT 2001
    * Fixed a bug with deeply nested callers.
    * Fixed nested uplevel() calls.
    - Forgot to include Test::More dependency
    - Added 5.006 dependency to Makefile.PL
    - Removed unnecessary die and warn overrides
    - Added DIRE WARNING to the docs.

0.05  Wed Sep 19 06:00:12 EDT 2001
    * Things were still pretty broken.  I *think* I have the tests
      sorted out now.
    * Blows over any CORE::GLOBAL::caller, die or warn you might have
      set.  I'll fix this soon.

0.04  Wed Sep 19 04:28:19 EDT 2001
    * Ooops, we'd broken caller().  Turns out the tests were wrong.

0.03  Wed Sep 19 03:41:59 EDT 2001
    * Greatly simplified the uplevel logic

0.02  Wed Sep 19 03:03:10 EDT 2001
    * Fools croak()

0.01  Wed Sep 19 00:19:38 EDT 2001
    * First working version
    * Fools caller(), die() and warn().
    - Needs more work against, say, Carp.
    - Needs more work to check that it doesn't break the
      subtleties of caller, die and warn.


t/lib/Test/More.pm
package Test::More;

use 5.004;

use strict;
use Carp;
use Test::Utils;

BEGIN {
    require Test::Simple;
    *TESTOUT = \*Test::Simple::TESTOUT;
    *TESTERR = \*Test::Simple::TESTERR;
}

require Exporter;
use vars qw($VERSION @ISA @EXPORT $TODO);
$VERSION = '0.20';
@ISA    = qw(Exporter);
@EXPORT = qw(ok use_ok require_ok
             is isnt like
             skip todo
             pass fail
             eq_array eq_hash eq_set
             $TODO
             plan
             can_ok  isa_ok
            );


sub import {
    my($class, $plan, @args) = @_;

    if( defined $plan ) {
        if( $plan eq 'skip_all' ) {
            $Test::Simple::Skip_All = 1;
            my $out = "1..0";
            $out .= " # Skip @args" if @args;
            $out .= "\n";

            my_print *TESTOUT, $out;
            exit(0);
        }
        else {
            Test::Simple->import($plan => @args);
        }
    }
    else {
        Test::Simple->import;
    }

    __PACKAGE__->_export_to_level(1, __PACKAGE__);
}

# 5.004's Exporter doesn't have export_to_level.
sub _export_to_level
{
      my $pkg = shift;
      my $level = shift;
      (undef) = shift;                  # XXX redundant arg
      my $callpkg = caller($level);
      $pkg->export($callpkg, @_);
}


=head1 NAME

Test::More - yet another framework for writing test scripts

=head1 SYNOPSIS

  use Test::More tests => $Num_Tests;
  # or
  use Test::More qw(no_plan);
  # or
  use Test::More skip_all => $reason;

  BEGIN { use_ok( 'Some::Module' ); }
  require_ok( 'Some::Module' );

  # Various ways to say "ok"
  ok($this eq $that, $test_name);

  is  ($this, $that,    $test_name);
  isnt($this, $that,    $test_name);
  like($this, qr/that/, $test_name);

  SKIP: {
      skip $why, $how_many unless $have_some_feature;

      ok( foo(),       $test_name );
      is( foo(42), 23, $test_name );
  };

  TODO: {
      local $TODO = $why;

      ok( foo(),       $test_name );
      is( foo(42), 23, $test_name );
  };

  can_ok($module, @methods);
  isa_ok($object, $class);

  pass($test_name);
  fail($test_name);

  # Utility comparison functions.
  eq_array(\@this, \@that);
  eq_hash(\%this, \%that);
  eq_set(\@this, \@that);

  # UNIMPLEMENTED!!!
  my @status = Test::More::status;

  # UNIMPLEMENTED!!!
  BAIL_OUT($why);


=head1 DESCRIPTION

If you're just getting started writing tests, have a look at
Test::Simple first.  This is a drop in replacement for Test::Simple
which you can switch to once you get the hang of basic testing.

This module provides a very wide range of testing utilities.  Various
ways to say "ok", facilities to skip tests, test future features
and compare complicated data structures.


=head2 I love it when a plan comes together

Before anything else, you need a testing plan.  This basically declares
how many tests your script is going to run to protect against premature
failure.

The prefered way to do this is to declare a plan when you C<use Test::More>.

  use Test::More tests => $Num_Tests;

There are rare cases when you will not know beforehand how many tests
your script is going to run.  In this case, you can declare that you
have no plan.  (Try to avoid using this as it weakens your test.)

  use Test::More qw(no_plan);

In some cases, you'll want to completely skip an entire testing script.

  use Test::More skip_all => $skip_reason;

Your script will declare a skip with the reason why you skipped and
exit immediately with a zero (success).  See L<Test::Harness> for
details.


=head2 Test names

By convention, each test is assigned a number in order.  This is
largely done automatically for you.  However, its often very useful to
assign a name to each test.  Which would you rather see:

  ok 4
  not ok 5
  ok 6

or

  ok 4 - basic multi-variable
  not ok 5 - simple exponential
  ok 6 - force == mass * acceleration

The later gives you some idea of what failed.  It also makes it easier
to find the test in your script, simply search for "simple
exponential".

All test functions take a name argument.  Its optional, but highly
suggested that you use it.


=head2 I'm ok, you're not ok.

The basic purpose of this module is to print out either "ok #" or "not
ok #" depending on if a given test succeeded or failed.  Everything
else is just gravy.

All of the following print "ok" or "not ok" depending on if the test
succeeded or failed.  They all also return true or false,
respectively.

=over 4

=item B<ok>

  ok($this eq $that, $test_name);

This simply evaluates any expression (C<$this eq $that> is just a
simple example) and uses that to determine if the test succeeded or
failed.  A true expression passes, a false one fails.  Very simple.

For example:

    ok( $exp{9} == 81,                   'simple exponential' );
    ok( Film->can('db_Main'),            'set_db()' );
    ok( $p->tests == 4,                  'saw tests' );
    ok( !grep !defined $_, @items,       'items populated' );

(Mnemonic:  "This is ok.")

$test_name is a very short description of the test that will be printed
out.  It makes it very easy to find a test in your script when it fails
and gives others an idea of your intentions.  $test_name is optional,
but we B<very> strongly encourage its use.

Should an ok() fail, it will produce some diagnostics:

    not ok 18 - sufficient mucus
    #     Failed test 18 (foo.t at line 42)

This is actually Test::Simple's ok() routine.

=cut

# We get ok() from Test::Simple's import().

=item B<is>

=item B<isnt>

  is  ( $this, $that, $test_name );
  isnt( $this, $that, $test_name );

Similar to ok(), is() and isnt() compare their two arguments
with C<eq> and C<ne> respectively and use the result of that to
determine if the test succeeded or failed.  So these:

    # Is the ultimate answer 42?
    is( ultimate_answer(), 42,          "Meaning of Life" );

    # $foo isn't empty
    isnt( $foo, '',     "Got some foo" );

are similar to these:

    ok( ultimate_answer() eq 42,        "Meaning of Life" );
    ok( $foo ne '',     "Got some foo" );

(Mnemonic:  "This is that."  "This isn't that.")

So why use these?  They produce better diagnostics on failure.  ok()
cannot know what you are testing for (beyond the name), but is() and
isnt() know what the test was and why it failed.  For example this
test:

    my $foo = 'waffle';  my $bar = 'yarblokos';
    is( $foo, $bar,   'Is foo the same as bar?' );

Will produce something like this:

    not ok 17 - Is foo the same as bar?
    #     Failed test 1 (foo.t at line 139)
    #          got: 'waffle'
    #     expected: 'yarblokos'

So you can figure out what went wrong without rerunning the test.

You are encouraged to use is() and isnt() over ok() where possible,
however do not be tempted to use them to find out if something is
true or false!

  # XXX BAD!  $pope->isa('Catholic') eq 1
  is( $pope->isa('Catholic'), 1,        'Is the Pope Catholic?' );

This does not check if C<$pope->isa('Catholic')> is true, it checks if
it returns 1.  Very different.  Similar caveats exist for false and 0.
In these cases, use ok().

  ok( $pope->isa('Catholic') ),         'Is the Pope Catholic?' );

For those grammatical pedants out there, there's an C<isn't()>
function which is an alias of isnt().

=cut

sub is ($$;$) {
    my($this, $that, $name) = @_;

    my $test;
    {
        local $^W = 0;   # so is(undef, undef) works quietly.
        $test = $this eq $that;
    }
    my $ok = @_ == 3 ? ok($test, $name)
                     : ok($test);

    unless( $ok ) {
        $this = defined $this ? "'$this'" : 'undef';
        $that = defined $that ? "'$that'" : 'undef';
        diagnostic *TESTERR, sprintf <<DIAGNOSTIC, $this, $that;
#          got: %s
#     expected: %s
DIAGNOSTIC

    }

    return $ok;
}

sub isnt ($$;$) {
    my($this, $that, $name) = @_;

    my $test;
    {
        local $^W = 0;   # so isnt(undef, undef) works quietly.
        $test = $this ne $that;
    }

    my $ok = @_ == 3 ? ok($test, $name)
                     : ok($test);

    unless( $ok ) {
        $that = defined $that ? "'$that'" : 'undef';

        diagnostic *TESTERR, sprintf <<DIAGNOSTIC, $that;
#     it should not be %s
#     but it is.
DIAGNOSTIC

    }

    return $ok;
}

*isn't = \&isnt;


=item B<like>

  like( $this, qr/that/, $test_name );

Similar to ok(), like() matches $this against the regex C<qr/that/>.

So this:

    like($this, qr/that/, 'this is like that');

is similar to:

    ok( $this =~ /that/, 'this is like that');

(Mnemonic "This is like that".)

The second argument is a regular expression.  It may be given as a
regex reference (ie. C<qr//>) or (for better compatibility with older
perls) as a string that looks like a regex (alternative delimiters are
currently not supported):

    like( $this, '/that/', 'this is like that' );

Regex options may be placed on the end (C<'/that/i'>).

Its advantages over ok() are similar to that of is() and isnt().  Better
diagnostics on failure.

=cut

sub like ($$;$) {
    my($this, $regex, $name) = @_;

    my $ok = 0;
    if( ref $regex eq 'Regexp' ) {
        local $^W = 0;
        $ok = @_ == 3 ? ok( $this =~ $regex ? 1 : 0, $name )
                      : ok( $this =~ $regex ? 1 : 0 );
    }
    # Check if it looks like '/foo/i'
    elsif( my($re, $opts) = $regex =~ m{^ /(.*)/ (\w*) $ }sx ) {
        local $^W = 0;
        $ok = @_ == 3 ? ok( $this =~ /(?$opts)$re/ ? 1 : 0, $name )
                      : ok( $this =~ /(?$opts)$re/ ? 1 : 0 );
    }
    else {
        # Can't use fail() here, the call stack will be fucked.
        my $ok = @_ == 3 ? ok(0, $name )
                         : ok(0);

        diagnostic *TESTERR, <<ERR;
#     '$regex' doesn't look much like a regex to me.  Failing the test.
ERR

        return $ok;
    }

    unless( $ok ) {
        $this = defined $this ? "'$this'" : 'undef';
        diagnostic *TESTERR, sprintf <<DIAGNOSTIC, $this;
#                   %s
#     doesn't match '$regex'
DIAGNOSTIC

    }

    return $ok;
}

=item B<can_ok>

  can_ok($module, @methods);
  can_ok($object, @methods);

Checks to make sure the $module or $object can do these @methods
(works with functions, too).

    can_ok('Foo', qw(this that whatever));

is almost exactly like saying:

    ok( Foo->can('this') && 
        Foo->can('that') && 
        Foo->can('whatever') 
      );

only without all the typing and with a better interface.  Handy for
quickly testing an interface.

=cut

sub can_ok ($@) {
    my($proto, @methods) = @_;
    my $class= ref $proto || $proto;

    my @nok = ();
    foreach my $method (@methods) {
        my $test = "$class->can('$method')";
        eval $test || push @nok, $method;
    }

    my $name;
    $name = @methods == 1 ? "$class->can($methods[0])" 
                          : "$class->can(...)";
    
    ok( !@nok, $name );

    diagnostic *TESTERR, map "#     $class->can('$_') failed\n", @nok;

    return !@nok;
}

=item B<isa_ok>

  isa_ok($object, $class);

Checks to see if the given $object->isa($class).  Also checks to make
sure the object was defined in the first place.  Handy for this sort
of thing:

    my $obj = Some::Module->new;
    isa_ok( $obj, 'Some::Module' );

where you'd otherwise have to write

    my $obj = Some::Module->new;
    ok( defined $obj && $obj->isa('Some::Module') );

to safeguard against your test script blowing up.

=cut

sub isa_ok ($$) {
    my($object, $class) = @_;

    my $diag;
    my $name = "object->isa('$class')";
    if( !defined $object ) {
        $diag = "The object isn't defined";
    }
    elsif( !ref $object ) {
        $diag = "The object isn't a reference";
    }
    elsif( !$object->isa($class) ) {
        $diag = "The object isn't a '$class'";
    }

    if( $diag ) {
        ok( 0, $name );
        diagnostic *TESTERR, "#     $diag\n";
        return 0;
    }
    else {
        ok( 1, $name );
        return 1;
    }
}


=item B<pass>

=item B<fail>

  pass($test_name);
  fail($test_name);

Sometimes you just want to say that the tests have passed.  Usually
the case is you've got some complicated condition that is difficult to
wedge into an ok().  In this case, you can simply use pass() (to
declare the test ok) or fail (for not ok).  They are synonyms for
ok(1) and ok(0).

Use these very, very, very sparingly.

=cut

sub pass (;$) {
    my($name) = @_;
    return @_ == 1 ? ok(1, $name)
                   : ok(1);
}

sub fail (;$) {
    my($name) = @_;
    return @_ == 1 ? ok(0, $name)
                   : ok(0);
}

=back

=head2 Module tests

You usually want to test if the module you're testing loads ok, rather
than just vomiting if its load fails.  For such purposes we have
C<use_ok> and C<require_ok>.

=over 4

=item B<use_ok>

   BEGIN { use_ok($module); }
   BEGIN { use_ok($module, @imports); }

These simply use the given $module and test to make sure the load
happened ok.  Its recommended that you run use_ok() inside a BEGIN
block so its functions are exported at compile-time and prototypes are
properly honored.

If @imports are given, they are passed through to the use.  So this:

   BEGIN { use_ok('Some::Module', qw(foo bar)) }

is like doing this:

   use Some::Module qw(foo bar);


=cut

sub use_ok ($;@) {
    my($module, @imports) = @_;
    @imports = () unless @imports;

    my $pack = caller;

    eval <<USE;
package $pack;
require $module;
$module->import(\@imports);
USE

    my $ok = ok( !$@, "use $module;" );

    unless( $ok ) {
        chomp $@;
        diagnostic *TESTERR, <<DIAGNOSTIC;
#     Tried to use '$module'.
#     Error:  $@
DIAGNOSTIC

    }

    return $ok;
}

=item B<require_ok>

   require_ok($module);

Like use_ok(), except it requires the $module.

=cut

sub require_ok ($) {
    my($module) = shift;

    my $pack = caller;

    eval <<REQUIRE;
package $pack;
require $module;
REQUIRE

    my $ok = ok( !$@, "require $module;" );

    unless( $ok ) {
        chomp $@;
        diagnostic *TESTERR, <<DIAGNOSTIC;
#     Tried to require '$module'.
#     Error:  $@
DIAGNOSTIC

    }

    return $ok;
}

=back

=head2 Conditional tests

B<WARNING!> The following describes an I<experimental> interface that
is subject to change B<WITHOUT NOTICE>!  Use at your peril.

Sometimes running a test under certain conditions will cause the
test script to die.  A certain function or method isn't implemented
(such as fork() on MacOS), some resource isn't available (like a 
net connection) or a module isn't available.  In these cases it's
necessary to skip tests, or declare that they are supposed to fail
but will work in the future (a todo test).

For more details on skip and todo tests see L<Test::Harness>.

The way Test::More handles this is with a named block.  Basically, a
block of tests which can be skipped over or made todo.  It's best if I
just show you...

=over 4

=item B<SKIP: BLOCK>

  SKIP: {
      skip $why, $how_many if $condition;

      ...normal testing code goes here...
  }

This declares a block of tests to skip, $how_many tests there are,
$why and under what $condition to skip them.  An example is the
easiest way to illustrate:

    SKIP: {
        skip "Pigs don't fly here", 2 unless Pigs->can('fly');

        my $pig = Pigs->new;
        $pig->takeoff;

        ok( $pig->altitude > 0,         'Pig is airborne' );
        ok( $pig->airspeed > 0,         '  and moving'    );
    }

If pigs cannot fly, the whole block of tests will be skipped
completely.  Test::More will output special ok's which Test::Harness
interprets as skipped tests.  Its important to include $how_many tests
are in the block so the total number of tests comes out right (unless
you're using C<no_plan>).

You'll typically use this when a feature is missing, like an optional
module is not installed or the operating system doesn't have some
feature (like fork() or symlinks) or maybe you need an Internet
connection and one isn't available.

=for _Future
See L</Why are skip and todo so weird?>

=cut

#'#
sub skip {
    my($why, $how_many) = @_;
    unless( $how_many >= 1 ) {
        # $how_many can only be avoided when no_plan is in use.
        carp "skip() needs to know \$how_many tests are in the block"
          if $Test::Simple::Planned_Tests;
        $how_many = 1;
    }

    for( 1..$how_many ) {
        Test::Simple::_skipped($why);
    }

    local $^W = 0;
    last SKIP;
}


=item B<TODO: BLOCK>

    TODO: {
        local $TODO = $why;

        ...normal testing code goes here...
    }

Declares a block of tests you expect to fail and $why.  Perhaps it's
because you haven't fixed a bug or haven't finished a new feature:

    TODO: {
        local $TODO = "URI::Geller not finished";

        my $card = "Eight of clubs";
        is( URI::Geller->your_card, $card, 'Is THIS your card?' );

        my $spoon;
        URI::Geller->bend_spoon;
        is( $spoon, 'bent',    "Spoon bending, that's original" );
    }

With a todo block, the tests inside are expected to fail.  Test::More
will run the tests normally, but print out special flags indicating
they are "todo".  Test::Harness will interpret failures as being ok.
Should anything succeed, it will report it as an unexpected success.

The nice part about todo tests, as opposed to simply commenting out a
block of tests, is it's like having a programatic todo list.  You know
how much work is left to be done, you're aware of what bugs there are,
and you'll know immediately when they're fixed.

Once a todo test starts succeeding, simply move it outside the block.
When the block is empty, delete it.


=back

=head2 Comparision functions

Not everything is a simple eq check or regex.  There are times you
need to see if two arrays are equivalent, for instance.  For these
instances, Test::More provides a handful of useful functions.

B<NOTE> These are NOT well-tested on circular references.  Nor am I
quite sure what will happen with filehandles.

=over 4

=item B<eq_array>

  eq_array(\@this, \@that);

Checks if two arrays are equivalent.  This is a deep check, so
multi-level structures are handled correctly.

=cut

#'#
sub eq_array  {
    my($a1, $a2) = @_;
    return 0 unless @$a1 == @$a2;
    return 1 if $a1 eq $a2;

    my $ok = 1;
    for (0..$#{$a1}) {
        my($e1,$e2) = ($a1->[$_], $a2->[$_]);
        $ok = _deep_check($e1,$e2);
        last unless $ok;
    }
    return $ok;
}

sub _deep_check {
    my($e1, $e2) = @_;
    my $ok = 0;

    my $eq;
    {
        # Quiet unintialized value warnings when comparing undefs.
        local $^W = 0; 

        if( $e1 eq $e2 ) {
            $ok = 1;
        }
        else {
            if( UNIVERSAL::isa($e1, 'ARRAY') and
                UNIVERSAL::isa($e2, 'ARRAY') )
            {
                $ok = eq_array($e1, $e2);
            }
            elsif( UNIVERSAL::isa($e1, 'HASH') and
                   UNIVERSAL::isa($e2, 'HASH') )
            {
                $ok = eq_hash($e1, $e2);
            }
            else {
                $ok = 0;
            }
        }
    }

    return $ok;
}


=item B<eq_hash>

  eq_hash(\%this, \%that);

Determines if the two hashes contain the same keys and values.  This
is a deep check.

=cut

sub eq_hash {
    my($a1, $a2) = @_;
    return 0 unless keys %$a1 == keys %$a2;
    return 1 if $a1 eq $a2;

    my $ok = 1;
    foreach my $k (keys %$a1) {
        my($e1, $e2) = ($a1->{$k}, $a2->{$k});
        $ok = _deep_check($e1, $e2);
        last unless $ok;
    }

    return $ok;
}

=item B<eq_set>

  eq_set(\@this, \@that);

Similar to eq_array(), except the order of the elements is B<not>
important.  This is a deep check, but the irrelevancy of order only
applies to the top level.

=cut

# We must make sure that references are treated neutrally.  It really
# doesn't matter how we sort them, as long as both arrays are sorted
# with the same algorithm.
sub _bogus_sort { local $^W = 0;  ref $a ? 0 : $a cmp $b }

sub eq_set  {
    my($a1, $a2) = @_;
    return 0 unless @$a1 == @$a2;

    # There's faster ways to do this, but this is easiest.
    return eq_array( [sort _bogus_sort @$a1], [sort _bogus_sort @$a2] );
}


=back

=head1 NOTES

Test::More is B<explicitly> tested all the way back to perl 5.004.

=head1 BUGS and CAVEATS

=over 4

=item Making your own ok()

This will not do what you mean:

    sub my_ok {
        ok( @_ );
    }

    my_ok( 2 + 2 == 5, 'Basic addition' );

since ok() takes it's arguments as scalars, it will see the length of
@_ (2) and always pass the test.  You want to do this instead:

    sub my_ok {
        ok( $_[0], $_[1] );
    }

The other functions act similiarly.

=item The eq_* family have some caveats.

=item Test::Harness upgrades

no_plan and todo depend on new Test::Harness features and fixes.  If
you're going to distribute tests that use no_plan your end-users will
have to upgrade Test::Harness to the latest one on CPAN.

If you simply depend on Test::More, it's own dependencies will cause a
Test::Harness upgrade.

=back

=head1 AUTHOR

Michael G Schwern E<lt>schwern@pobox.comE<gt> with much inspiration from
Joshua Pritikin's Test module and lots of discussion with Barrie
Slaymaker and the perl-qa gang.


=head1 HISTORY

This is a case of convergent evolution with Joshua Pritikin's Test
module.  I was largely unware of its existence when I'd first
written my own ok() routines.  This module exists because I can't
figure out how to easily wedge test names into Test's interface (along
with a few other problems).

The goal here is to have a testing utility that's simple to learn,
quick to use and difficult to trip yourself up with while still
providing more flexibility than the existing Test.pm.  As such, the
names of the most common routines are kept tiny, special cases and
magic side-effects are kept to a minimum.  WYSIWYG.


=head1 SEE ALSO

L<Test::Simple> if all this confuses you and you just want to write
some tests.  You can upgrade to Test::More later (its forward
compatible).

L<Test> for a similar testing module.

L<Test::Harness> for details on how your test results are interpreted
by Perl.

L<Test::Unit> describes a very featureful unit testing interface.

L<Pod::Tests> shows the idea of embedded testing.

L<SelfTest> is another approach to embedded testing.

=cut

1;


t/lib/Test/Utils.pm
package Test::Utils;

use 5.004;

use strict;
require Exporter;
use vars qw($VERSION @EXPORT @EXPORT_TAGS @ISA);

$VERSION = '0.03';

@ISA = qw(Exporter);
@EXPORT = qw( my_print print diagnostic );
                


# Special print function to guard against $\ and -l munging.
sub my_print (*@) {
    my($fh, @args) = @_;

    local $\;
    print $fh @args;
}

sub diagnostic (*@) {
    my($fh, @args) = @_;

    # Escape each line with a #.
    foreach (@args) {
        s/^([^#])/#$1/;
        s/\n([^#])/\n#$1/g;
    }
    local $\;
    print $fh @args;
}

sub print { die "DON'T USE PRINT!  Use _print instead" }

1;


Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "apparently run a function in a higher stack frame",
  "AUTHOR" => "Michael Schwern <mschwern\@cpan.org>, David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Sub-Uplevel",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Sub::Uplevel",
  "PREREQ_PM" => {
    "Carp" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "0.2800",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => 0,
  "constant" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
